net/utils/net_bufpool.c

```diff
@@ -0,0 +1,172 @@
+/****************************************************************************
+ * net/utils/net_bufpool.c
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <nuttx/kmalloc.h>
+#include <nuttx/net/net.h>
+#include <nuttx/semaphore.h>
+
+#include "utils/utils.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* The node to store in the pool */
+
+struct net_bufnode_s
+{
+  sq_entry_t node;
+};
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: net_bufpool_init
+ *
+ * Description:
+ *   Initialize a network buffer pool.
+ *
+ * Input Parameters:
+ *   pool - The pool to be initialized
+ *
+ ****************************************************************************/
+
+void net_bufpool_init(FAR struct net_bufpool_s *pool)
+{
+  int i;
+  unsigned int maxalloc;
+
+  if (pool->dynalloc > 0)
+    {
+      maxalloc = pool->u.maxalloc > 0 ? pool->u.maxalloc : INT16_MAX;
+    }
+  else
+    {
+      maxalloc = pool->prealloc;
+    }
+
+  nxsem_init(&pool->u.sem, 0, maxalloc);
+
+  sq_init(&pool->freebuffers);
+  for (i = 0; i < pool->prealloc; i++)
```

Comment:
can call net_bufpool_init in net_bufpool_timedalloc automatically and change net_bufpool_timedalloc to static